api/librdb-api.h

```diff
@@ -379,6 +383,33 @@ _LIBRDB_API  size_t RDB_bulkLen(RdbParser *p, RdbBulk b);
 
 int RDB_isRefBulk(RdbParser *p, RdbBulk b);
 
+/****************************************************************
+ * Multiple levels registration
+ * Some of the more advanced configuration might require parsing different data
+ * types at different levels of the parser.
+ *
+ * The callbacks that are common to all levels and not related to current key
+ * parsing (lookup HANDLERS_COMMON_CALLBACKS), if registered at different levels
+ * then all of them will be called, one by one, starting from level 0.
+ *
+ * As for the callbacks of RDB object types, each level has its own way to
+ * handle the data with distinct set of callbacks interfaces. In case of multiple
+ * levels registration, the application should configure for each RDB object type
+ * at what level it is needed to get parsed by calling `RDB_handleByLevel()`.
+ * Otherwise, the parser will resolve it by parsing and calling handlers that are
+ * registered at lowest level.
+ *****************************************************************/
+typedef enum RdbObjType {
+    /* values aligned with RDB opcodes (defines.h) */
+    RDB_OBJ_TYPE_STRING=0,
+    RDB_OBJ_TYPE_QUICKLIST=14,
+    RDB_OBJ_TYPE_QUICKLIST2=18,
+    RDB_OBJ_TYPE_MAX=22
+} RdbObjType;
```

Comment:
Maybe instead of enumerating all over again and spamming the API with redundant long enum, as an alternative, delete the enum and  function `RDB_handleByLevel` will be declared with `int opcode` instead of `RdbObjType` and a comment will direct (the advanced user) to lookup file `defines.h` for opcode values. The terminology should to be close to RDB `opcodes`. To whom this feature is relevant, it will make a lot of sense.

Reply:
not sure if it'll make it cleaner, it'll remove type checking and will mean that anyone using it has to re-define these values explicitly.

in any case, i think the comment would still apply:
1. maybe the user doesn't want to set it per opcode, but rather per data type, maybe with an `encoded` or `packed` boolean.
2. maybe we better declare the API for this library to support all types (even if we don't implement parsing yet), it could help us realize where we're going and make API design decisions.

Reply:
Another option is before the parser starts to parse, it analysis what are the handlers that are registered at each level and it will configure based on it to prefer the finer callbacks (higher levels). 

It can be a nice solution except that what happens if at level 2, for example, the app uses a built-in **filter** with all the wrapping callbacks (including handleString), and in addition another set of customized callbacks that don’t have handleString`. It won’t be right to prefer level 2 in this case just because of the built-in filter has `handleString` callback. ... So maybe the parser will prefer a higher level if all the registered handlers at that level have `handleString` callback. Otherwise it will lookup for lower levels.

Reply:
> not sure if it'll make it cleaner, it'll remove type checking and will mean that anyone using it has to re-define these values explicitly.
> 
> in any case, i think the comment would still apply:
> 
> 1. maybe the user doesn't want to set it per opcode, but rather per data type, maybe with an `encoded` or `packed` boolean.
> 2. maybe we better declare the API for this library to support all types (even if we don't implement parsing yet), it could help us realize where we're going and make API design decisions.

Ok, your recommendation is better. My last suggestion is too stricts and discards the option to configure during parsing based on rdb version that being read.

So if i understand correctly you recommend to have a limited set of data-types enum of {string,list,set,zset,hash,module,stream}. I wouldn't rush to use 'encoded' as long as there is no real need to distinct it. We can reserve `flags` in the API for future use.

Reply:
Yes, but let's design the api for the entire thing, it'll be easier to reason with things.

Reply:
Ok. I will make it on a distinct PR. Thanks!